climada/engine/impact.py

```diff
@@ -390,6 +388,47 @@ def impact_per_year(self, all_years=True, year_range=None):
             year_set[year] = sum(self.at_event[orig_year == year])
         return year_set
 
+    def impact_at_reg(self, agg_regions=None):
+        """Aggregate impact on given aggregation regions. This method works
+        only if Impact.imp_mat was stored during the impact calculation.
+
+        Parameters
+        ----------
+        agg_regions : np.array, list, pd.Series (optional)
+            It has a lenght equal to the number of coordinates points in exposure
+            and it reports what macro-regions these points belong to. For example,
+            asuming there are three centroids and agg_regions = ['A', 'A', 'B']
+            then impact of the first and second centroids will be assigned to
+            region A, whereas impact from the second centroid will be assigned
+            to area B. If no aggregation regions are passed, the method aggregates
+            impact at the country (admin_0) level.
+            Default is None.
+
+        Returns
+        -------
+        pd.DataFrame
+            Contains the aggregated data per event.
+            Rows: Hzard events. Columns: Aggregation regions.
+        """
+        if self.imp_mat.nnz == 0:
+            raise ValueError("The aggregated impact cannot be computed as no"
+                             "Impact.imp_mat was stored during the impact calculation")
+
+        if agg_regions is None:
+            agg_regions = pd.Series(u_coord.get_country_code(self.coord_exp[:,0],
+                                                             self.coord_exp[:,1]))
+        elif not isinstance(agg_regions, pd.Series):
+            agg_regions = pd.Series(agg_regions)
```

Comment:
Why make it a pandas series instead of a numpy array?

Reply:
for two reasons:
- I find it cleaner when as you now what impact belongs to what region simply by looking at the dataframe
- I like the `pd.Series.unique()` method because it retains the order. Doing the same with `numpy` is more cumbersome, to my knowledge.

Reply:
I would argue that this is a misuse of a pandas series. 

`
  I find it cleaner as you know what impact belongs to what region simply by looking at the dataframe
`
That is independent on using numpy arrays or pandas series. But using a pandas series when it is not needed is imho actually less clean.
`
  I like the pd.Series.unique() method because it retains the order. Doing the same with numpy is more cumbersome, to my knowledge.
`
I would argue that this is suboptimal as an important point is now hidden in the subtle internal working of the `pandas.series.unique()` vs the `np.unique()` methods.

@peanutfun : what do you think?

Reply:
- on the first point you are right: having a `pd.Series` is not strictly needed to generate a `pd.DataFrame`

- on the second: I am transforming `agg_regions` into `pd.Series` only when the user supplies it as an `np.array' or `list` . But the user can supply a `pd.Series` in the first place. So I don't think it's a misuse, I just wrote the code in a way that it works with a `pd.Series`. 
So I think your doubt is rather: should we use `pandas` for things that can be done with `numpy`? Not sure I have a clear answer to that, as this would probably apply to any code that make use of `pandas`.

Reply:
The point is that `pd.series` are objects that are made to be used in pandas dataframes, but they are not made to be used a standalone objects. They are essentially numpy arrays, with some more. Thus, instead of making an unclear use of the `pd.series`, I suggest to use numpy arrays, and to make the ordering requirement clear (as currently this is a hidden feature).

Reply:
I don't really see how the order is important because the order in the `agg_regions` parameter is determined by the "order" of the exposure points. Also, the order of columns in the final data frame does not appear relevant to me 😅

I would actually argue that the unordered result of `pandas.Series.unique` does not need further explanation, whereas I feel we should add a note to the docstring that the unique items will be ordered when using `np.unique`. So, I think the easiest solution is to just stick to the current implementation.

@chahank: Following our discussion on the "clean" indexing, I just realized that `pandas.Series.unique` returns a numpy array, so everything is fine from that perspective.

Reply:
You are right, the order does not count. I am now using only `numpy` arrays.

Reply:
> I would argue that this is a misuse of a pandas series.

Not necessarily. According to the docs a `Series` is a "one-dimensional `ndarray` with axis labels (including time series)". Sounds useful also outside of a `DataFrame`.
🤷

Reply:
I very much agree with @emanuel-schmid here 😃

I'm fine with the current implementation. However, please make sure to only call `np.unique` **once**, and store the result for later use. It is currently executed twice and will be costly for large arrays. I can also take care of that when merging. Are we ready for that?